src/shortestpaths/bellman-ford.jl

```diff
@@ -116,13 +116,40 @@ to all other vertices. In addition, `enumerate_paths(state, v, d)` will return
 a vector representing the path from vertex `v` to vertex `d`.
 """
 function enumerate_paths(state::AbstractPathState, vs::AbstractVector{<:Integer})
-    parents = state.parents
-    T = eltype(parents)
+    T = eltype(state.parents)
+    all_paths = Vector{T}[Vector{eltype(state.parents)}() for _ in 1:length(vs)]
+    return enumerate_paths!(all_paths, state, vs)
+end
+enumerate_paths(state::AbstractPathState, v::Integer) = enumerate_paths(state, v:v)[1]
+function enumerate_paths(state::AbstractPathState)
+    return enumerate_paths(state, 1:length(state.parents))
+end
+
+"""
+    enumerate_paths!(paths::AbstractVector{<:AbstractVector}, state, vs::AbstractVector{Int})
+
+In-place version of [`enumerate_paths`](@ref).
+
+`paths` must be a `Vector{Vectors{eltype(state.parents)}}`
 
+`enumerate_paths!` should be more efficient when used in a loop,
+as the same memory can be used for each iteration.
+"""
+function enumerate_paths!(
+    all_paths::AbstractVector{<:AbstractVector},
+    state::AbstractPathState,
+    vs::AbstractVector{<:Integer},
+)
+    Base.require_one_based_indexing(all_paths)
+    Base.require_one_based_indexing(vs)
+    @assert length(all_paths) == length(vs)
```

Comment:
From the documentation:

> Warning
>  
> An assert might be disabled at some optimization levels.
  Assert should therefore only be used as a debugging tool and
  not used for authentication verification (e.g., verifying
  passwords or checking array bounds). The code must not rely
  on the side effects of running cond for the correct behavior
  of a function.

Reply:
Ok this is now an ArgumentError instead, and I've tested expliticly calling `enumerate_paths!` with both correct and incorrect length inputs so its also tested